subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/DefaultIvyArtifactSet.java

```diff
@@ -38,11 +37,24 @@ public class DefaultIvyArtifactSet extends DefaultDomainObjectSet<IvyArtifact> i
     private final FileCollection files;
     private final NotationParser<Object, IvyArtifact> ivyArtifactParser;
 
-    public DefaultIvyArtifactSet(String publicationName, NotationParser<Object, IvyArtifact> ivyArtifactParser, FileCollectionFactory fileCollectionFactory, CollectionCallbackActionDecorator collectionCallbackActionDecorator) {
+    public DefaultIvyArtifactSet(
+        String publicationName,
+        NotationParser<Object, IvyArtifact> ivyArtifactParser,
+        FileCollectionFactory fileCollectionFactory,
+        CollectionCallbackActionDecorator collectionCallbackActionDecorator,
+        TaskDependencyFactory taskDependencyFactory
+    ) {
         super(IvyArtifact.class, collectionCallbackActionDecorator);
         this.publicationName = publicationName;
         this.ivyArtifactParser = ivyArtifactParser;
-        this.files = fileCollectionFactory.create(new ArtifactsTaskDependency(), new ArtifactsFileCollection());
+        this.files = fileCollectionFactory.create(
```

Comment:
It seems all call sites now pass a `visitingDependencies` instance. It looks like an opportunity to introduce something a `FileCollectionFactory#create(Consumer<? super TaskDependencyResolveContext> visitDependencies, MinimalFileSet contents)` and avoid the contagious dependency on `taskDependencyFactory` in some places.

Reply:
There was [one call site](https://github.com/gradle/gradle/blob/03f99e890509b20cc47d3986177dd49f95a23b99/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultPublishArtifactSet.java#L57) that intentionally stored the `TaskDependency` produced by the factory, and I left it as it was. The other three got simpler with the change you proposed 👍

✅ in https://github.com/gradle/gradle/pull/22433/commits/716a4332d0680d14f5fda86efd39d26610ee1a67